File: ai_research/dataset_generation/domain/assertion.go
package domain

import (
	"encoding/json"
	"regexp"
	"strings"
)

type Assertion struct {
	AssertionText   string
	CodeObjectNames []string
}

func parseAssertion(assertionText string) Assertion {
	re := regexp.MustCompile(`\{(?P<object>[^}]*)}`)

	trimmedAssertionText := strings.Trim(assertionText, `"`)

	matches := re.FindAllStringSubmatch(assertionText, -1)
	if matches != nil {
		codeObjectNames := make([]string, len(matches))
		for i, match := range matches {
			codeObjectNames[i] = match[re.SubexpIndex("object")]
		}
		return Assertion{
			AssertionText:   trimmedAssertionText,
			CodeObjectNames: codeObjectNames,
		}
	}

	return Assertion{
		AssertionText:   trimmedAssertionText,
		CodeObjectNames: []string{},
	}
}

func (p Assertion) MarshalJSON() ([]byte, error) {
	return json.Marshal(map[string]interface{}{
		"assertionText":   p.AssertionText,
		"codeObjectNames": p.CodeObjectNames,
	})
}

File: ai_research/dataset_generation/domain/assertion.py
import re
from dataclasses import dataclass


@dataclass
class Assertion:
    assertion_text: str
    code_object_names: list[str]

    def to_dict(self) -> dict[str, object]:
        return {
            "assertionText": self.assertion_text,
            "codeObjectNames": self.code_object_names,
        }


def parse_assertion(assertion_text: str) -> Assertion:
    matches = re.findall(r"\{([^}]*)}", assertion_text)
    return Assertion(
        assertion_text=assertion_text.strip('"'),
        code_object_names=matches,
    )
